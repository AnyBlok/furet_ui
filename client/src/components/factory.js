import Vue from 'vue';

const components = {};
const componentNames = [];

export const defineComponent = (elementName, declaration) => {
  let klass;
  if (elementName in components) {
    klass = components[elementName];
  } else {
    klass = {
      template: undefined,
      extend: [],
      prototypes: [],
    };
    components[elementName] = klass;
    componentNames.push(elementName);
  }
  if (declaration.template !== undefined) klass.template = declaration.template;
  if (declaration.prototype !== undefined) {
    klass.prototypes.push(declaration.prototype);
  }
  if (declaration.extend !== undefined) {
    declaration.extend.forEach((extend) => {
      klass.extend.push(extend);
    });
  }
};

window.defineComponent = defineComponent;

const getPrototypeFor = (elementName) => {
  const component = components[elementName];
  const prototypes = [];
  component.extend.forEach((otherElementName) => {
    prototypes.push(...getPrototypeFor(otherElementName));
  });
  prototypes.push(...component.prototypes);
  return prototypes;
};

export const createComponents = () => {
  componentNames.forEach((elementName) => {
    const component = components[elementName];
    const mixins = getPrototypeFor(elementName);
    Vue.component(elementName, (resolve) => {
      resolve({
        template: component.template,
        mixins,
      });
    });
  });
};
