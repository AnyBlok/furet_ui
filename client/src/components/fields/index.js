import './vue2-editor';
