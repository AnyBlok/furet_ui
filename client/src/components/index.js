import './mixins';
import './app';
import './homepage';
import './login';
import './logout';
import './about';
import './list';

export { createComponents, default } from './factory';
