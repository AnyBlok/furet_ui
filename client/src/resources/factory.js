import _ from 'underscore';

const resources = {};
export const resourceNames = [];

export const removeResource = (resourceName) => {
  const index = resourceNames.indexOf(resourceName);
  if (index !== -1) {
    delete resources[resourceName];
    resourceNames.splice(index, 1);
  }
};

export const defineResource = (resourceName, declaration) => {
  let klass;
  if (resourceName in resources) {
    klass = resources[resourceName];
  } else {
    klass = {
      children: [],
      name: resourceName,
      path: resourceName,
      templateName: resourceName,
      props: [],
      mustBeAuthenticated: false,
    };
    resources[resourceName] = klass;
    resourceNames.push(resourceName);
  }
  if (declaration.path !== undefined) klass.path = declaration.path;
  if (declaration.templateName !== undefined) klass.templateName = declaration.templateName;
  if (declaration.props !== undefined) klass.props = declaration.props;
  if (declaration.mustBeAuthenticated !== undefined) {
    klass.mustBeAuthenticated = declaration.mustBeAuthenticated;
  }
  if (declaration.children !== undefined) {
    declaration.children.forEach((child) => {
      let childDefinition = _.find(klass.children, c => c.name === child.name);
      if (childDefinition === undefined) {
        childDefinition = {
          name: child.name,
          completeName: `${resourceName}_${child.name}`,
          templateName: child.name,
          path: child.name,
          props: [],
          mustBeAuthenticated: false,
        };
        klass.children.push(childDefinition);
      }
      if (child.path !== undefined) childDefinition.path = child.path;
      if (child.templateName !== undefined) childDefinition.templateName = child.templateName;
      if (child.props !== undefined) childDefinition.props.push(...child.props);
      if (child.mustBeAuthenticated !== undefined) {
        childDefinition.mustBeAuthenticated = child.mustBeAuthenticated;
      }
    });
  }
};

window.defineResource = defineResource;

const createRoute = (resource, defaultTemplateProps) => {
  let path = `/${resource.path}`;
  let templateProps = defaultTemplateProps || '';
  if (resource.path === resource.name) {
    if (resource.props.length !== 0) {
      resource.props.forEach((prop) => {
        path += `/:${prop}`;
        templateProps += ` v-bind:${prop}="${prop}"`;
      });
    }
  } else if (resource.props.length !== 0) {
    resource.props.forEach((prop) => {
      templateProps += ` v-bind:${prop}="${prop}"`;
    });
  }
  const route = {
    name: resource.completeName || resource.name,
    path,
    props: templateProps !== '',
    meta: { requiresAuth: resource.mustBeAuthenticated },
    component: { props: resource.props },
  };
  if (resource.children === undefined || resource.children.length === 0) {
    route.component.template = `<${resource.templateName}${templateProps} />`;
  } else {
    route.children = [];
    if (resource.templateName === resource.name) {
      route.component.template = '<router-view />';
    } else {
      route.component.template = `<${resource.templateName}${templateProps} />`;
    }
    resource.children.forEach((child) => {
      route.children.push(createRoute(child, templateProps));
    });
  }
  return route;
};

export const getRoutes = () => {
  const routes = [];
  resourceNames.forEach((resourceName) => {
    const resource = resources[resourceName];
    routes.push(createRoute(resource));
  });
  return routes;
};
