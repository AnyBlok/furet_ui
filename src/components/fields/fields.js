export const fields = {
  list: {},
  form: {},
  thumbnail: {},
  kanban: {},
};
export default {
    fields,
}
