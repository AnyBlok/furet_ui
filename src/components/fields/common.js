/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import {defineComponent} from '../factory'
import { debounce } from 'debounce';


/**
 * Eval a value (that could be anything String / integer / null / undefined / bool)
 * and cast as a boolean or null.
 *
 * @param {any} value - A value to cast to boolean
 * @returns {Boolean?} return one of those values [true, false, null]
 *
 * Have a look to tests/unit/fields/boolean.spec.js for a complet list of
 * use case.
 */
export const safe_eval_boolean = value => {
  if (value === null || value === undefined) {
    return null;
  }

  // if dev use new String("") or new Boolean(true) this give an object so
  // we convert to its primitive type
  if (
    typeof value === "object" &&
    ["[object String]", "[object Boolean]"].includes(toString.call(value))
  ) {
    value = value.valueOf();
  }

  if (typeof value === "string") {
    if (
      value.trim().toLowerCase() === "null" ||
      value.trim().toLowerCase() === "undefined"
    ) {
      return null;
    }
    return ["true", "t", "yes", "y", "on", "1"].includes(
      value.trim().toLowerCase()
    );
  }

  if (typeof value === "number") {
    return value === 1;
  }

  if (typeof value === "boolean") {
    return value;
  }

  return false;
};

export const safe_eval = (condition, fields, resource) => {
    const now = Date.now(),
          toDate = (v) => new Date(v),
          selectors = Object.assign((resource.manager || {}).selectors || {}, resource.selectors || {}),
          tabs = resource.tabs;
    fields  // lint
    selectors  // lint
    tabs  // lint
    now  // lint
    toDate  // lint
    let res = false;
    try {
        res = eval(condition) ? true : false;
    } catch (e) {
      console.log(e)
    }
    return res;
}

export const listTemplate = `
  <div>
    <span v-if="isHidden" />
    <span v-else>{{value}}</span>
  </div>
`


defineComponent('furet-ui-list-field-common', {
  prototype: {
    props: ['resource', 'data', 'config'],
    computed: {
      value () {
        return this.data[this.config.name] || '';
      },
      isHidden () {
        if (this.config.hidden === undefined) return false;
        return safe_eval(this.config.hidden, this.data || {}, this.resource);
      },
    },
  }
});


defineComponent('furet-ui-form-field-common-tooltip', {
  template: `
    <div v-if="isHidden" />
    <b-tooltip 
      v-bind:label="getTooltip" 
      v-bind:position="tooltipPosition"
      v-bind:style="{'width': '100%'}"
      v-else
    >
      <slot />
    </b-tooltip>`,
  prototype: {
    props: ['resource', 'data', 'config'],
    computed: {
      isHidden () {
        if (this.config.hidden === undefined) return false;
        return safe_eval(this.config.hidden, this.data || {}, this.resource);
      },
      getTooltip () {
        return this.config.tooltip || '';
      },
      tooltipPosition () {
        return this.config.tooltip_position || 'is-top';
      },
    },
  },
})


defineComponent('furet-ui-form-field-common-tooltip-field', {
  template: `
    <furet-ui-form-field-common-tooltip
      v-bind:resource="resource"
      v-bind:data="data"
      v-bind:config="config"
    >
      <b-field 
        v-bind:label="config.label"
        v-bind:type="getType"
        v-bind:message="getMessage"
        v-bind:style="{'width': 'inherit'}"
      >
        <slot />
      </b-field>
    </furet-ui-form-field-common-tooltip>`,
  prototype: {
    props: ['resource', 'data', 'config'],
    computed: {
      value () {
        return this.data && this.data[this.config.name] || '';
      },
      isRequired () {
        return safe_eval(this.config.required, this.data || {}, this.resource);
      },
      getType () {
        if (this.isRequired) {
          if (this.value) return 'is-info';
          return 'is-danger';
        }
        return '';
      },
      getMessage () {
        if (this.isRequired) {
          if (!this.value) return this.$i18n.t('fields.common.required');
        }
        return ''
      },
    },
  },
})


defineComponent('furet-ui-form-field-common-readonly', {
  prototype: {
    props: ['resource', 'data', 'config'],
    inject: ['partIsReadonly'],
    computed: {
      isReadonly () {
        if (this.resource.readonly) return true;
        if (this.partIsReadonly()) return true;
        const readonlyParams = safe_eval(this.config.readonly, this.data || {}, this.resource);
        if (this.config.writable) {
          const writableParams = safe_eval(this.config.writable, this.data || {}, this.resource);
          return readonlyParams && ! writableParams
        }
        return readonlyParams;
      },
    },
  },
})


defineComponent('furet-ui-form-field-common', {
  extend: ['furet-ui-form-field-common-readonly'],
  prototype: {
    inject: ['updateChangeState'],
    computed: {
      value () {
        return this.data && this.data[this.config.name] || null;
      },
    },
    methods: {
      updateValue: debounce(function(value, merge) {
        const action = {
          model: this.resource.model,
          pk: this.resource.pks,
          uuid: this.resource.uuid,
          fieldname: this.config.name,
          value,
          merge
        }
        this.updateChangeState(action);
      }, 200),
    },
  }
});
