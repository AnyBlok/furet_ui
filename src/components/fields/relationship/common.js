/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import axios from "axios";
import { defineComponent } from "../../factory";
import { pk2string } from "../../../store/modules/data";
import { debounce } from "debounce";

const safe_eval = (style, fields) => {
  fields; // lint
  return eval(style);
};

/**
 * furet-ui-field-relationship component is a mixin used to manage relationship to add
 * helper:
 *
 *   * format: function to display an entry of the relationship
 *   * openResource: strategie to open the resource as the main resource
 *   * getKey: transform primary key to string
 *   * getStyle: return a style string for ``a`` tag
 *
 * added injection:
 *   * getEntry
 *   * pushInBreadcrumb
 *
 * @mixin
 *
 * @param {Object} config - A config object to manage the behaviour of the component
 * @param {Object} data - An object that contains data to display. The key to use
 *                        in set in the `config.key`
 * @param {Object} resource - A resource object used to properly bind data with parents
 *                            tags and manage reactivity.
 *
 * ``config`` Object contains
 * @param {String} name - the key to use in the ``data`` object where is store the value
 * @param {String} model - A model name, needed to display the data
 * @param {String} display - An evaluate string to display the entry
 */
defineComponent("furet-ui-field-relationship", {
  prototype: {
    inject: ["getEntry", "pushInBreadcrumb"],
    methods: {
      format(display, fields) {
        return safe_eval(display, fields);
      },
      getKey(value) {
        return pk2string(value.pk);
      },
      getStyle(value) {
        let style = "";
        if (this.config.style !== undefined) {
          const data = this.getEntry(this.config.model, value.pk);
          style = safe_eval(this.config.style, data, {});
        }
        return style;
      },
      openResource(value) {
        const params = { code: this.$route.params.code, menuId: 0, id: 0 };
        if (this.config.menu) params.menuId = this.config.menu;
        if (this.config.resource) {
          const query = { mode: "form", pks: JSON.stringify(value) };
          params.id = this.config.resource;
          this.pushInBreadcrumb();
          this.$router.push({ name: "resource", params, query });
        }
      },
    },
  },
});

export const RelationShipX2MList = `
  <div>
    <span v-if="isHidden" />
    <div v-else>
      <span 
        v-for="value in values"
        class="tag" 
        v-bind:key="getKey(value)"
        v-bind:style="getStyle(value)"
      >
        <a v-on:click.stop="openResource(value)">{{value.label}}</a>
      </span>
    </div>
  </div>`;

defineComponent("furet-ui-field-relationship-search", {
  prototype: {
    data() {
      return {
        pks: [],
      };
    },
    computed: {
      choices() {
        const res = [];
        this.pks.forEach((pk) => {
          res.push({
            pk,
            label: this.format(
              this.config.display,
              this.getEntry(this.config.model, pk)
            ),
          });
        });
        return res;
      },
    },
    methods: {
      beforeOnChange() {
        /** Method to overwrite in order to
         * clear current value before searching things.
         */
      },
      onChange: debounce(function(value) {
        this._onChange(value);
      }, 200),
      _onChange(value) {
        this.beforeOnChange();
        const params = {
          "context[model]": this.config.model,
          "context[fields]": this.config.fields.toString(),
          limit: this.config.limit,
        };
        this.config.filter_by.forEach((filter) => {
          params[`filter[${filter}][ilike]`] = value;
        });
        if (this.value && Array.isArray(this.value)) {
          // TODO: We needs to improuve in case of composite key
          const first_key = this.config.remote_columns[0];
          const key_values = this.value
            .map((key) => {
              return key[first_key];
            })
            .join(",");
          if (key_values) params[`~filter[${first_key}][in]`] = key_values;
        }
        axios
          .get(`/furet-ui/resource/${this.resource.id}/crud`, { params })
          .then((response) => {
            this.$dispatchAll(response.data.data);
            this.pks = response.data.pks;
            this.loading = false;
          });
      }
    },
  },
});
