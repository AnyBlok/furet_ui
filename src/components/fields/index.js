import { fields } from './fields';
import './common';
import './boolean';
import './datetime';
import './integer';
import './password';
import './timestamp';
import './selection';
import './string';

import './relationship';

import { defineComponent } from '../factory';

defineComponent('furet-ui-unknown-field', {
  template : `<div>Unknow Field <strong>{{config.type}}</strong></div>`,
  prototype: {
    props: ['config', 'data'],
  }
})

defineComponent('furet-ui-field', {
  template : `
    <component 
      v-bind:is="component" 
      v-bind:resource="resource" 
      v-bind:config="config" 
      v-bind:data="data"
    />
  `,
  prototype: {
    props: ['resource', 'config', 'data'],
    computed: {
      component () {
        return (fields[this.resource.type] || {})[this.config.type] || 'furet-ui-unknown-field'
      },
    },
  }
})
