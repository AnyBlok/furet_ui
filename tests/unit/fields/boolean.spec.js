import { mount } from "@vue/test-utils";
import { getComponentPrototype } from "@/components/factory";
import { safe_eval_boolean } from "@/components/fields/common";

const localVue = global.localVue;
const store = global.store;
const provide = {
  partIsReadonly: () => {
    return false;
  },
  updateChangeState: () => {}
};

describe("Field.Boolean for Resource.List", () => {
  const ListBooleanField = getComponentPrototype("furet-ui-list-field-boolean");

  it("Empty", () => {
    const wrapper = mount(ListBooleanField, {
      store,
      localVue,
      provide,
      propsData: {
        resource: {},
        data: {},
        config: {}
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("With true value", () => {
    const wrapper = mount(ListBooleanField, {
      store,
      localVue,
      provide,
      propsData: {
        resource: {},
        config: {
          name: "fieldName"
        },
        data: {
          fieldName: "true"
        }
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("With false value", () => {
    const wrapper = mount(ListBooleanField, {
      store,
      localVue,
      provide,
      propsData: {
        resource: {},
        config: {
          name: "fieldName"
        },
        data: {
          fieldName: "false"
        }
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("With security issue value", () => {
    const wrapper = mount(ListBooleanField, {
      store,
      localVue,
      provide,
      propsData: {
        resource: {},
        config: {
          name: "fieldName"
        },
        data: {
          fieldName: "throw Error('this is a security issue')"
        }
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});

describe("Field.Boolean for Resource.Form", () => {
  const FormBooleanField = getComponentPrototype("furet-ui-form-field-boolean");

  it("Empty", () => {
    const wrapper = mount(FormBooleanField, {
      store,
      localVue,
      provide,
      propsData: {
        resource: {},
        data: {},
        config: {}
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});

describe("Test eval boolean", () => {
  it("test true values", () => {
    const TrulyValues = [
      [true, "true"],
      [1, "1"],
      ["1", "'1'"],
      ["True", "'True'"],
      ["true", "'true'"],
      ["t", "'t'"],
      ["yes", "'yes'"],
      ["y", "'y'"],
      ["on", "'on'"],
      [new String("1"), "new String('1')"],
      [new String("TrUe"), "new String('TrUe')"],
      [String("1"), "String('1')"],
      [String("true"), "String('true')"],
      [Boolean(true), "Boolean(true)"],
      [new Boolean(true), "new Boolean(true)"]
    ];
    TrulyValues.forEach(val => {
      expect(
        safe_eval_boolean(val[0]),
        "We expected safe_eval_boolean( `" + val[1] + "` ) to be true"
      ).toBe(true);
    });
  });
  it("test false values", () => {
    const FalsyValues = [
      [false, "false"],
      [-1, "-1"],
      [0, "0"],
      ["-1", '"-1"'],
      ["0", '"0"'],
      ["False", '"False"'],
      ["false", '"false"'],
      ["no", '"no"'],
      ["n", '"n"'],
      ["f", '"f"'],
      [12, "12"],
      ["12", '"12"'],
      [new String("-1"), 'new String("-1")'],
      [new String("0"), 'new String("0")'],
      [new String("12"), 'new String("12")'],
      [String("-1"), 'String("-1")'],
      [String("12"), 'String("12")'],
      [String("0"), 'String("0")'],
      [String("false"), 'String("false")'],
      [String("False"), 'String("False")'],
      [new Boolean(false), "new Boolean(false)"],
      [Boolean(false), "Boolean(false)"]
    ];
    FalsyValues.forEach(val => {
      expect(
        safe_eval_boolean(val[0]),
        "We expected safe_eval_boolean(`" + val[1] + "`) to be false"
      ).toBe(false);
    });
  });
  it("test null values", () => {
    const NullValues = [
      [null, "null"],
      ["null", '"null"'],
      [String("null"), 'String("null")'],
      [new String("null"), 'new String("null")'],
      [undefined, "undefined"],
      ["undefined", '"undefined"'],
      [String("undefined"), 'String("undefined")'],
      [new String("undefined"), 'new String("undefined")']
    ];
    NullValues.forEach(val => {
      expect(
        safe_eval_boolean(val[0]),
        "We expected safe_eval_boolean( `" + val[1] + "` ) to be null"
      ).toBe(null);
    });
  });
});
