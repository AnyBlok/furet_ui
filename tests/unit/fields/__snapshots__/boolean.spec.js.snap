// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field.Boolean for Resource.Form Empty 1`] = `
<span
  class="is-primary is-top is-medium b-tooltip"
  data-label=""
  style="transition-delay: 0ms; width: 100%;"
>
  <label
    class="b-checkbox checkbox"
  >
    <input
      true-value="true"
      type="checkbox"
      value="false"
    />
     
    <span
      class="check"
    />
     
    <span
      class="control-label"
    >
      
        
      
    </span>
  </label>
</span>
`;

exports[`Field.Boolean for Resource.Form With security issue value 1`] = `
<span
  class="is-primary is-top is-medium b-tooltip"
  data-label=""
  style="transition-delay: 0ms; width: 100%;"
>
  <label
    class="b-checkbox checkbox"
  >
    <input
      true-value="true"
      type="checkbox"
      value="false"
    />
     
    <span
      class="check"
    />
     
    <span
      class="control-label"
    >
      
        
      
    </span>
  </label>
</span>
`;

exports[`Field.Boolean for Resource.Form With value and label 1`] = `
<span
  class="is-primary is-top is-medium b-tooltip"
  data-label=""
  style="transition-delay: 0ms; width: 100%;"
>
  <label
    class="b-checkbox checkbox is-disabled"
    disabled="disabled"
  >
    <input
      disabled="disabled"
      true-value="true"
      type="checkbox"
      value="false"
    />
     
    <span
      class="check"
    />
     
    <span
      class="control-label"
    >
      
        My field label
      
    </span>
  </label>
</span>
`;

exports[`Field.Boolean for Resource.Form With value, label and icon 1`] = `
<span
  class="is-primary is-top is-medium b-tooltip"
  data-label=""
  style="transition-delay: 0ms; width: 100%;"
>
  <label
    class="b-checkbox checkbox"
  >
    <input
      true-value="true"
      type="checkbox"
      value="false"
    />
     
    <span
      class="check"
    />
     
    <span
      class="control-label"
    >
      
        My field label
      
    </span>
  </label>
</span>
`;

exports[`Field.Boolean for Resource.List Empty 1`] = `
<div>
  <label
    class="b-checkbox checkbox is-disabled"
    disabled="disabled"
  >
    <input
      disabled="disabled"
      true-value="true"
      type="checkbox"
      value="false"
    />
     
    <span
      class="check"
    />
     
    <span
      class="control-label"
    />
  </label>
</div>
`;

exports[`Field.Boolean for Resource.List With false value 1`] = `
<div>
  <label
    class="b-checkbox checkbox is-disabled"
    disabled="disabled"
  >
    <input
      disabled="disabled"
      true-value="true"
      type="checkbox"
      value="false"
    />
     
    <span
      class="check"
    />
     
    <span
      class="control-label"
    />
  </label>
</div>
`;

exports[`Field.Boolean for Resource.List With security issue value 1`] = `
<div>
  <label
    class="b-checkbox checkbox is-disabled"
    disabled="disabled"
  >
    <input
      disabled="disabled"
      true-value="true"
      type="checkbox"
      value="false"
    />
     
    <span
      class="check"
    />
     
    <span
      class="control-label"
    />
  </label>
</div>
`;

exports[`Field.Boolean for Resource.List With true value 1`] = `
<div>
  <label
    class="b-checkbox checkbox is-disabled"
    disabled="disabled"
  >
    <input
      disabled="disabled"
      true-value="true"
      type="checkbox"
      value="false"
    />
     
    <span
      class="check"
    />
     
    <span
      class="control-label"
    />
  </label>
</div>
`;

exports[`Field.YesNo YesNo for Resource.List Empty 1`] = `
<div>
  <span>
    No
  </span>
</div>
`;

exports[`Field.YesNo YesNo for Resource.List With false value 1`] = `
<div>
  <span>
    No
  </span>
</div>
`;

exports[`Field.YesNo YesNo for Resource.List With security issue value 1`] = `
<div>
  <span>
    No
  </span>
</div>
`;

exports[`Field.YesNo YesNo for Resource.List With true value 1`] = `
<div>
  <span>
    Yes
  </span>
</div>
`;
